### PR TITLE
Feature/dynamic flag struct size retrieval

### PR DESCRIPTION
**Pull request by cakeless**

### CODE REVIEW

File: vmicore/src/os/windows/KernelAccess.cpp [hunk 1]
```diff
@@ -168,23 +168,60 @@ uint32_t KernelAccess::extractFlagValue32(addr_t flagBaseVA, size_t startBit, si
     return getFlagValue(flagValue, startBit, endBit);
 }
 
+uint64_t KernelAccess::extractFlagValue64(addr_t flagBaseVA, size_t startBit, size_t endBit) const
+{
+    expectSaneKernelAddress(flagBaseVA, __func__); // NOLINT(cppcoreguidelines-pro-bounds-array-to-pointer-decay)
```
Comment (rageagainsthepc) [inline vmicore/src/os/windows/KernelAccess.cpp:173]:
```suggestion
    expectSaneKernelAddress(flagBaseVA, static_cast<const char*>(__func__));
```

File: vmicore/src/os/windows/KernelAccess.cpp [hunk 2]
```diff
@@ -168,23 +168,60 @@ uint32_t KernelAccess::extractFlagValue32(addr_t flagBaseVA, size_t startBit, si
...
-                              kernelOffsets.mmvadFlags.protection.endBit);
+    auto flagStructSize = vmiInterface->getStructSizeFromJson(kernelOffsets.mmvadFlags.structName);
+    switch (flagStructSize)
+    {
+        // NOLINTNEXTLINE(cppcoreguidelines-avoid-magic-numbers)
+        case 4:
+            return extractFlagValue32(getMmVadShortFlagsAddr(vadShortBaseVA),
+                                      kernelOffsets.mmvadFlags.protection.startBit,
+                                      kernelOffsets.mmvadFlags.protection.endBit);
+        // As of now, there are 32 Protectionvalues
```
Comment (rageagainsthepc) [inline vmicore/src/os/windows/KernelAccess.cpp:194]: nit: You could add an assert to verify this.

File: vmicore/src/os/windows/KernelAccess.cpp [hunk 3]
```diff
@@ -168,23 +168,60 @@ uint32_t KernelAccess::extractFlagValue32(addr_t flagBaseVA, size_t startBit, si
...
+        // NOLINTNEXTLINE(cppcoreguidelines-avoid-magic-numbers)
+        case 8:
+            return static_cast<uint32_t>(extractFlagValue64(getMmVadShortFlagsAddr(vadShortBaseVA),
+                                                            kernelOffsets.mmvadFlags.protection.startBit,
+                                                            kernelOffsets.mmvadFlags.protection.endBit));
+        default:
+            throw VmiException(std::string(kernelOffsets.mmvadFlags.structName) + ": " +
```
Comment (rageagainsthepc) [inline vmicore/src/os/windows/KernelAccess.cpp:201]: Use `fmt::format()`

File: vmicore/src/os/windows/KernelAccess.cpp [hunk 4]
```diff
@@ -168,23 +168,60 @@ uint32_t KernelAccess::extractFlagValue32(addr_t flagBaseVA, size_t startBit, si
...
 {
-    return static_cast<bool>(extractFlagValue32(getMmVadShortFlagsAddr(vadShortBaseVA),
-                                                kernelOffsets.mmvadFlags.privateMemory.startBit,
-                                                kernelOffsets.mmvadFlags.privateMemory.endBit));
+    auto flagStructSize = vmiInterface->getStructSizeFromJson(kernelOffsets.mmvadFlags.structName);
+    switch (flagStructSize)
+    {
+        // NOLINTNEXTLINE(cppcoreguidelines-avoid-magic-numbers)
+        case 4:
+            return static_cast<bool>(extractFlagValue32(getMmVadShortFlagsAddr(vadShortBaseVA),
```
Comment (rageagainsthepc) [inline vmicore/src/os/windows/KernelAccess.cpp:213]: nit: Same here. Add an assert statement in order to verify that this is actually a single flag.

File: vmicore/src/os/windows/KernelAccess.cpp [hunk 5]
```diff
@@ -194,23 +237,71 @@ addr_t KernelAccess::getMmSectionFlagsAddr(addr_t controlAreaBaseVA) const
...
-    return static_cast<bool>(extractFlagValue32(getMmSectionFlagsAddr(controlAreaBaseVA),
-                                                kernelOffsets.mmsectionFlags.beingDeleted.startBit,
-                                                kernelOffsets.mmsectionFlags.beingDeleted.endBit));
+    auto flagStructSize = vmiInterface->getStructSizeFromJson(KernelStructOffsets::mmsection_flags::structName);
+    assert((kernelOffsets.mmsectionFlags.beingDeleted.endBit - kernelOffsets.mmsectionFlags.beingDeleted.startBit) ==
+           1);
+    switch (flagStructSize)
+    {
+        // NOLINTNEXTLINE(cppcoreguidelines-avoid-magic-numbers)
+        case 4:
```
Comment (rageagainsthepc) [inline vmicore/src/os/windows/KernelAccess.cpp:246]: why not:
```suggestion
        case sizeof(uint32_t):
```

File: vmicore/src/os/windows/KernelAccess.cpp [hunk 6]
```diff
@@ -194,23 +237,71 @@ addr_t KernelAccess::getMmSectionFlagsAddr(addr_t controlAreaBaseVA) const
...
 }
 
 bool KernelAccess::extractIsFile(addr_t controlAreaBaseVA) const
 {
-    return static_cast<bool>(extractFlagValue32(getMmSectionFlagsAddr(controlAreaBaseVA),
-                                                kernelOffsets.mmsectionFlags.file.startBit,
-                                                kernelOffsets.mmsectionFlags.file.endBit));
+    auto flagStructSize = vmiInterface->getStructSizeFromJson(kernelOffsets.mmsectionFlags.structName);
+    assert((kernelOffsets.mmsectionFlags.file.endBit - kernelOffsets.mmsectionFlags.file.startBit) == 1);
+    switch (flagStructSize)
```
Comment (rageagainsthepc) [inline vmicore/src/os/windows/KernelAccess.cpp:288]: That's a lot of code duplication. Why not extract this to a new function?

File: vmicore/src/os/windows/KernelAccess.cpp (hunk 5 above)
Comment (cakeless) [reply vmicore/src/os/windows/KernelAccess.cpp:246]: In my opinion '4' has better readability

**Review by rageagainsthepc — APPROVED** [review]

**Pull request merged**